2-WebApp-graph-user/2-5-HybridFlow/Program.cs

```diff
@@ -0,0 +1,73 @@
+using Microsoft.AspNetCore.Authentication.OpenIdConnect;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Mvc.Authorization;
+using Microsoft.Identity.Web;
+using WebApp_OpenIDConnect_DotNet.Services;
+
+var serviceCollection = new ServiceCollection();
+serviceCollection.AddSingleton<IConfidentialClientApplicationService, ConfidentialClientApplicationService>();
+
+var serviceProvider = serviceCollection.BuildServiceProvider();
+
+var builder = WebApplication.CreateBuilder(args);
+
+builder.Services.Add(serviceCollection.Single());
+
+builder.Services.Concat(serviceCollection);
+
+builder.Services.AddAuthentication(OpenIdConnectDefaults.AuthenticationScheme)
+    .AddMicrosoftIdentityWebApp(options =>
+    {
+        options.ResponseType = "code";
+        options.Events.OnAuthorizationCodeReceived = async context =>
+        {
+            context.TokenEndpointRequest.Parameters.TryGetValue("code_verifier", out var codeVerifier);
+            context.HandleCodeRedemption();
+
+            var clientService = serviceProvider.GetService<IConfidentialClientApplicationService>();
+
+            var authResult = await clientService.GetAuthenticationResultAsync(new String[] { "user.read", "mail.read" }, context.ProtocolMessage.Code, codeVerifier);
+            var spaCode = authResult.SpaAuthCode;
+            context.Response.Cookies.Append("Microsoft.Identity.Hybrid.Authentication.Cookie", spaCode);
+
+            context.TokenEndpointResponse.AccessToken = authResult.AccessToken;
+            context.TokenEndpointResponse.IdToken = authResult.IdToken;
+
+        };
+        builder.Configuration.Bind("AzureAd", options);
+    });
```

Comment:
This is the work horse of the app.

It handles the authentication and retrieval of the Auth code from Azure while placing it in a cookie.

This does use PKCE. I've been reading a lot of chatter about whether or not PKCE is overkill in a confidential app but the consensus seems to be 'it's a good idea'.

There are some issues with the commented 'out of the box' methods.

E.g. 

`EnableTokenAcquisitionToCallDownstreamApi`
`AddMicrosoftGraph`
`AddInMemoryTokenCaches`


The problem is that if those methods are used the library seems to redeem the Auth code automatically making it unusable in the client.

I can dive in and make a deeper diagnosis later.

For clarity, the lines below:

```
                context.TokenEndpointResponse.AccessToken = authResult.AccessToken;
                context.TokenEndpointResponse.IdToken = authResult.IdToken;
```

Are needed to prevent an exception being thrown by the next step of the authentication flow.